devito/data.py

```diff
@@ -122,6 +126,24 @@ def reset(self):
         """
         self[:] = 0.0
 
+    def __reduce_ex__(self, proto):
+        """ Pickling support."""
+        return type(self), self.__getnewargs__(), self.__getstate__()
+
+    def __getnewargs__(self):
+        return (self.shape, self.dimensions, self.dtype)
```

Comment:
so pickling requires **all** objects used to initialise a given object? 

I wonder whether we could rather provide (not sure if it's a pickling convention or I'm just inventing it now) a new method in each Pickable class, namely ``__new_frompickling__``, to be called by ``__new__`` depending on the arguments in input to ``__new__`` itself (``__new__``). Just a random thought. In general, it feels weird that we have to keep the entire input only for pickling...